File: frontend/src/features/admin/services/fare-config-service.ts
import axiosClient from "@/services/http/axios";
import { FareConfig, FareConfigRequest } from "../types";

interface ApiListResponse<T> {
    code: number;
    message: string;
    result: T[];
}

interface ApiSingleResponse<T> {
    code: number;
    message: string;
    result: T;
}

export const fareConfigService = {
    /** Lấy tất cả giá vé đang active */
    getAll: async (): Promise<FareConfig[]> => {
        const { data } = await axiosClient.get<ApiListResponse<FareConfig>>("/pricing/fares");
        return data.result;
    },

    /** Lấy giá vé active theo tuyến + loại xe */
    getActiveFare: async (routeId: number, busTypeId: number, date?: string): Promise<FareConfig> => {
        const params: Record<string, string | number> = { routeId, busTypeId };
        if (date) params.date = date;
        const { data } = await axiosClient.get<ApiSingleResponse<FareConfig>>("/pricing/fares/active", { params });
        return data.result;
    },

    /** Tạo mới hoặc cập nhật giá (SCD Type 2) */
    upsert: async (request: FareConfigRequest): Promise<FareConfig> => {
        const { data } = await axiosClient.post<ApiSingleResponse<FareConfig>>("/pricing/fares/upsert", request);
        return data.result;
    },
};

File: backend/src/main/java/com/bus/system/modules/pricing/service/FareConfigService.java
package com.bus.system.modules.pricing.service;

import com.bus.system.modules.pricing.dto.request.FareConfigRequest;
import com.bus.system.modules.pricing.dto.response.FareConfigResponse;

import java.time.LocalDate;
import java.util.List;

public interface FareConfigService {
    // Tạo mới hoặc cập nhật giá (Logic SCD Type 2)
    FareConfigResponse upsertFare(FareConfigRequest request);

    // Lấy giá đang áp dụng (Active)
    FareConfigResponse getActiveFare(Long routeId, Long busTypeId, LocalDate date);

    // Lấy tất cả giá vé đang áp dụng
    List<FareConfigResponse> getAllActiveFares();
}